debugProtocol.json

```diff
@@ -2326,6 +2338,133 @@
+			"properties": {
+				"memoryReference": {
+					"type": "string",
+					"description": "Memory reference to the base location from which data should be read."
+				},
+				"offset": {
+					"type": "integer",
+					"description": "Optional offset (in bytes) to be applied to the reference location before reading data.  Can be negative."
+				},
+				"count": {
+					"type": "integer",
+					"description": "Number of bytes to read at the specified location and offset."
+				}
+			},
+			"required": [ "memoryReference", "count" ]
+		},
+		"ReadMemoryResponse": {
+			"allOf": [ { "$ref": "#/definitions/Response" }, {
+				"type": "object",
+				"description": "Response to 'readMemory' request.",
+				"properties": {
+					"body": {
+						"type": "object",
+						"properties": {
+							"address": {
+								"type": "string",
+								"description": "The address of the first byte of data returned.  Treated as a hex value if prefixed with '0x', or as a decimal value otherwise."
+							},
+							"unreadableBytes": {
+								"type": "integer",
+								"description": "The number of unreadable bytes encountered after the last successfully read byte.  This can be used to determine the number of bytes that must be skipped before a subsequent 'readMemory' request will succeed."
+							},
+							"data": {
+								"type": "string",
+								"description": "The bytes read from memory, encoded using base64."
+							}
+						},
+						"required": [ "address" ]
+					}
+				}
+			}]
+		},
+
+		"DisassembleRequest": {
+			"allOf": [ { "$ref": "#/definitions/Request" }, {
+				"type": "object",
+				"description": "Disassembles code stored at the provided location.",
+				"properties": {
+					"command": {
+						"type": "string",
+						"enum": [ "disassemble" ]
+					},
+					"arguments": {
+						"$ref": "#/definitions/DisassembleArguments"
+					}
+				},
+				"required": [ "command", "arguments" ]
+			}]
+		},
+		"DisassembleArguments": {
+			"type": "object",
+			"description": "Arguments for 'disassemble' request.",
+			"properties": {
+				"memoryReference": {
+					"type": "string",
+					"description": "Memory reference to the base location containing the instructions to disassemble."
+				},
+				"offset": {
+					"type": "integer",
+					"description": "Optional offset (in bytes) to be applied to the reference location before disassembling.  Can be negative."
+				},
+				"instructionOffset": {
+					"type": "integer",
+					"description": "Optional offset (in instructions) to be applied after the byte offset (if any) before disassembling.  Can be negative."
+				},
+				"instructionCount": {
+					"type": "integer",
+					"description": "Number of instructions to disassemble starting at the specified location and offset.  An adapter must return exactly this number of instructions - any unavailable instructions should be replaced with an implementation-defined 'invalid instruction' value."
```

Comment:
Does `instructionCount` translate well with variable instruction lengths? I mean it is hard to ask or deliver a count. Both Intel (up to 15 bytes) and ARM have variable lengths? I see issues with both developing a DA and a DA's client. Should this be more like a memory read request? Specify the number of bytes to decode. Like @andrewcrawley found out, it is a lot trickier than one would imagine :-) There are alignment issues as well. Once we prototype, we will know more.

Reply:
Usually, a disassembly request with a symbol/section reference always succeeds. Normally, we make that request with what we find on the stack trace. 

I wish I could have a 10 mins phone conversation with someone on this team :-)

Reply:
Some times there can be data embedded along with instructions. Usually at the beginning or the end of a function.

Reply:
I've uploaded a file to show how data can be interspersed with instructions...I just took a random file I had, disassembled it and pasted a few functions.

https://github.com/haneefdm/cortex-debug-samples/blob/master/demo/tmp.s

Reply:
The "lazy scrolling" UI knows that it needs to request N lines of disassembly to fill in after scrolling, so it issues a request saying "starting at <`memoryReference` + `offset`>, provide the next N instructions".  It's up to the DA to handle variable-length instructions by continuing to disassemble until it has the requested number of instructions.

The only string field on a `DisassembledInstruction` that is interpreted in any way by the UI is the `address` field.  This means that you can return anything you want in the other fields and the disassembly UI will just display them as-is, so it's fine to use ".word	0x00c1b280" or whatever as the `instruction` value if a memory location is actually known to contain data.

Here's an example.  I compiled the following code:

```C++
int main()
{
	start:
	std::cout << "Hello World!\n"; 
	goto start;
}
```

In the VS disassembler, the `DisassembledInstructions` are mapped as follows:

![image](https://user-images.githubusercontent.com/17841637/58441149-60db2300-8094-11e9-968a-e36565e6d2db.png)

Reply:
@andrewcrawley  Yes nice, so the instruction can be data as well --  `.word` type thing. I was strictly thinking of 'instructions'. And, I think, get 'N' instructions as part of a request would also work fine. And the instruction count includes the `.word` type stuff? To me, the term `instruction` is not precise but I can't offer an alternative. The proposal says it **must return exactly this number of instructions**

My assumption (that you confirmed) is that if the `memoryReference + offset` is bogus, so will the output.  It is the DA clients responsibility to make a proper request. The DA will make no attempt to figure out if the request is proper or adjust.

I do think negative offsets are super useful for implementing lazy query/rendering. But how does the DA client make such a request -- a valid one. Can I just request a 1000 instructions before the `<memoryReference + offset>` and then from there the `instructionOffset` applies? But why have two types of offsets? What if the DA client asks for 1000 instructions before but there aren't that many out there? I think this is a lot of responsibility on the DA client to form proper requests

I am thinking `offset` is not needed assuming your heuristic works esp. for negative values for instructions and the client can easily do the `<memoryReference + offset>` calculation anyway.

If available, I would love to look at your vsdbg implementation. Then everything might make sense.

Finally, isn't this a holiday for you?

Reply:
A `DisassembledInstruction` is basically equivalent to a line of text in the disassembly UI - it represents data decoded at a specific address, whether that is an instruction, data, or even inaccessible (such as an unmapped page).  If the host asks for 1000 instructions (in either direction), it's up to the DA to return 1000 `DisassembledInstruction` objects.  If there are only 100 instructions after the current location, and then an unmapped page of memory, the DA should return 100 valid instructions, then 900 "invalid" instructions, which must contain a valid decoded address (generally incremented [or decremented, if using negative offsets] from the last instruction address by whatever the minimum instruction size for the architecture is, taking any required alignment into account), but otherwise can contain whatever text you want to use to represent an invalid location.  Most engines in VS just use "??" for all fields.

The reason for having both a byte and instruction offset is because the VS disassembly API allows consumers to say "go to this address (represented by `memoryReference` + `offset`), skip `instructionOffset` instructions, then decode the next `instructionCount` instructions", and the disassembly window makes use of this by remembering the address of the last instruction decoded.  A UI could also work entirely by `instructionOffset`, in which case the byte `offset` wouldn't be necessary, but that's not how VS works.

vsdbg is not open source, unfortunately, but I MIEngine uses a similar heuristic - take a look at `SeekBack` in https://github.com/microsoft/MIEngine/blob/master/src/MIDebugEngine/Engine.Impl/Disassembly.cs

Yes, today is a holiday, but I'm happy to have a distraction from cleaning out my garage : )

Reply:
@andrewcrawley Thank you, sir, for your patience. I was not aware of what vsdbg was currently doing.